src/Accelerators/NNPA/Conversion/ONNXToZHigh/ONNXToZHighCommon.hpp

```diff
@@ -14,31 +14,44 @@
 
 #pragma once
 
+#include "llvm/ADT/STLExtras.h"
+
 #include "src/Accelerators/NNPA/Conversion/ONNXToZHigh/NNPALimit.h"
 #include "src/Accelerators/NNPA/Conversion/ONNXToZHigh/ONNXLegalityCheck.hpp"
 #include "src/Accelerators/NNPA/Support/LayoutHelper.hpp"
 #include "src/Dialect/ONNX/ONNXDimAnalysis.hpp"
 
+namespace onnx_mlir {
+
+const std::string CPU_DEVICE = "cpu";
+const std::string NNPA_DEVICE = "nnpa";
+
 template <typename OP_TYPE>
 void addDynamicallyLegalOpFor(mlir::ConversionTarget *target,
     const onnx_mlir::DimAnalysis *dimAnalysis,
-    mlir::ArrayRef<std::string> execNodesOnCpu) {
-  target->addDynamicallyLegalOp<OP_TYPE>([dimAnalysis, execNodesOnCpu](
+    llvm::function_ref<bool(OP_TYPE, const DimAnalysis *)> checkLegalityFn =
+        nullptr) {
+  target->addDynamicallyLegalOp<OP_TYPE>([dimAnalysis, checkLegalityFn](
                                              OP_TYPE op) {
     // Check operations to be forced to run on CPU.
     mlir::Operation *genericOp = op.getOperation();
-    mlir::StringAttr nodeName =
-        genericOp->getAttrOfType<mlir::StringAttr>("onnx_node_name");
-    if (nodeName) {
-      bool exists =
-          llvm::any_of(execNodesOnCpu, [nodeName](llvm::StringRef val) {
-            return nodeName.getValue().equals_insensitive(val);
-          });
-      if (exists)
-        return true;
-    }
+    mlir::StringAttr device =
+        genericOp->getAttrOfType<mlir::StringAttr>("device");
+    // If device is CPU, force to run the op on CPU.
+    if (device && device.getValue().equals_insensitive(CPU_DEVICE))
+      return true;
+    // If device is NNPA, force to run the op on NNPA.
+    if (device && device.getValue().equals_insensitive(NNPA_DEVICE))
```

Comment:
Do you mean something like `isNNPA() && isLegality()`? `device=NNPA` can be `forcing to NNPA` or `maybe good for NNPA`. I am OK going with one of them.

`Forcing to NNPA` is convenient when we annotate an op with `device=NNPA` directly and we really want that op go to NNPA despite of compiler optimizations.

`maybe good for NNPA` is safe when we use a cost model, since the cost model may have a mistake in assigning an op to NNPA (e.g. that op is not suitable for NNPA)

Reply:
`Forcing to NNPA` is also useful when we have dynamic shape, and we want an op to run NNPA because the compiler is not able to know if it is suitable for CPU or NNPA.

Reply:
I should have been clearer
```
assert(isLegal(xxx) && "trying to force an op to NNPA that is not perceived as legal for NNPA");
```